Java 21 SpringBoot 3 Migration

RiccardoPompili opened the pull request:
[pn-service-desk] - Aggiornamento ms

RiccardoPompili committed:
PN-16205 - story

RiccardoPompili committed:
PN-17484 - [pn-service-desk] - Aggiornamento ms

RiccardoPompili committed:
PN-17484 - [pn-service-desk] - Aggiornamento ms 2

RiccardoPompili committed:
PN-17484 - [pn-service-desk] - Aggiornamento ms 3

michelescara commented on docs/openapi/schemas-pn-service-desk-v1.yaml:
```diff
@@ -337,6 +343,7 @@ components:
     Denomination:
       description: Denominazione ente o persona fisica / ragione sociale
       type: string
+      format: password # NO EXTERNAL # NO EXTERNAL
```
Correggere la duplicazione del commento

michelescara commented on src/main/java/it/pagopa/pn/service/desk/exception/RestExceptionHandler.java:
```diff
@@ -13,6 +13,7 @@
 import reactor.core.publisher.Mono;
 
 import java.time.Instant;
+import java.util.Date;
```
Import non utilizzato, da rimuovere

michelescara commented on pom.xml:
```diff
@@ -47,26 +59,27 @@
-			<version>1.9.0</version>
+			<version>2.9.0-SPRINGBOOT3</version>
 		</dependency>
+		<!--
```
Rimuovere codice commentato

michelescara commented on pom.xml:
```diff
@@ -171,28 +184,30 @@
-					</execution>
-				</executions>
-			</plugin>
+			<!-->
```
Rimuovere codice commentato

michelescara commented on pom.xml:
```diff
@@ -203,22 +218,26 @@
 							<generateApiTests>false</generateApiTests>
 							<generateModelTests>false</generateModelTests>
 							<configOptions>
+								<useSpringBoot3>true</useSpringBoot3>
```
`useSpringBoot3` va usato solo per la generazione delle componenti server, mentre `useJakartaEe` per la generazione dei client. Correggere anche sotto.

michelescara commented on pom.xml:
```diff
@@ -203,22 +218,26 @@
 								<apiPackage>${project.groupId}.service.desk.generated.openapi.server.v1.api</apiPackage>
 								<configPackage>${project.groupId}.service.desk.generated.openapi.server.v1.config</configPackage>
-								<dateLibrary>java8</dateLibrary>
+								<dateLibrary>java11</dateLibrary>
```
Se non è necessario aggiornarla, la dateLibrary va lasciata così com'è. Correggere anche sotto.

michelescara commented on src/main/java/it/pagopa/pn/service/desk/mapper/InfoPaMapper.java:
```diff
@@ -6,6 +6,7 @@
 import it.pagopa.pn.service.desk.generated.openapi.server.v1.dto.SearchNotificationsResponse;
 
 import java.util.ArrayList;
+import java.util.Date;
```
Import non utilizzato, da rimuovere.

michelescara commented on src/main/java/it/pagopa/pn/service/desk/mapper/ProfileMapper.java:
```diff
@@ -8,6 +8,7 @@
 
 import java.time.Instant;
 import java.util.ArrayList;
+import java.util.Date;
```
Import non utilizzato, da rimuovere.

michelescara commented on src/main/java/it/pagopa/pn/service/desk/middleware/queue/consumer/handler/ExternalChannelEventHandler.java:
```diff
@@ -21,18 +21,16 @@ public ExternalChannelEventHandler(ExternalChannelResponseHandler externalChanne
-                throw ex;
-            }
-        };
+    @SqsListener(value = "${pn.service-desk.topics.internal-queue}", acknowledgementMode = SqsListenerAcknowledgementMode.ALWAYS)
```
Il riferimento alla coda è sbagliato. Questa coda raccoglie le response di ExternalChannel. Il riferimento corretto è `pn.service-desk.topics.externalchannel-queue`.

michelescara commented on src/main/java/it/pagopa/pn/service/desk/middleware/queue/consumer/handler/PaperChannelEventHandler.java:
```diff
@@ -1,36 +1,34 @@
-            } catch (Exception ex) {
-                log.error("Error in pnPaperChannelInboundConsumer {}", ex.getMessage());
-                throw ex;
+    @SqsListener(value = "${pn.service-desk.topics.internal-queue}", acknowledgementMode = SqsListenerAcknowledgementMode.ALWAYS)
```
Anche qui il riferimento alla coda è sbagliato.  Il riferimento corretto è `pn.service-desk.topics.paperchannel-queue`.

michelescara commented on src/main/java/it/pagopa/pn/service/desk/middleware/queue/consumer/handler/SafeStorageEventHandler.java:
```diff
@@ -1,42 +1,40 @@
-            }
-        };
+
+    @SqsListener(value = "${pn.service-desk.topics.internal-queue}", acknowledgementMode = SqsListenerAcknowledgementMode.ALWAYS)
```
Anche qui riferimento errato, correggere con `pn.service-desk.topics.safe-storage-events`.

michelescara commented on src/main/java/it/pagopa/pn/service/desk/middleware/queue/consumer/PnEventInboundService.java:
Questo service va rimosso.

michelescara commented on src/test/java/it/pagopa/pn/service/desk/service/impl/NotificationAndMessageServiceImplTest.java:
```diff
@@ -79,8 +79,9 @@ void searchNotificationsFromTaxIdWhenDataVaultClientEmpty(){
+        Assertions.assertNotNull(actualResponse);
         Assertions.assertNull(actualResponse.getMoreResult());
-        Assertions.assertNull(actualResponse.getNextPagesKey());
+        Assertions.assertTrue(actualResponse.getNextPagesKey().isEmpty());
```
Qui stiamo cambiando il comportamento del test, come mai?

michelescara commented on src/test/java/it/pagopa/pn/service/desk/middleware/db/dao/impl/AddressDAOImplTest.java:
```diff
@@ -4,12 +4,13 @@
 import it.pagopa.pn.service.desk.middleware.db.dao.AddressDAO;
 import it.pagopa.pn.service.desk.middleware.entities.PnServiceDeskAddress;
-import it.pagopa.pn.service.desk.middleware.entities.PnServiceDeskOperations;
+//import it.pagopa.pn.service.desk.middleware.entities.PnServiceDeskOperations;
```
Rimuovere commenti sugli import.

michelescara commented on pom.xml:
```diff
@@ -4,30 +4,42 @@
 	</parent>
 	<artifactId>pn-service-desk</artifactId>
-	<version>1.10.0-SNAPSHOT</version>
+	<version>1.10.0-SPRINGBOOT3</version>
```
@masciale89 qui non mi torna rispetto a quello che abbiamo fatto su [pn-emd-integration](https://github.com/pagopa/pn-emd-integration/pull/86). Che versione deve avere il microservizio?

RiccardoPompili committed:
PN-17484 - Correzioni PR

michelescara committed:
PN-17484: Restored java8 date library for compatibility.

michelescara committed:
PN-17484: Fix handlers. Add tests.

michelescara committed:
PN-17484: resolved copilot reviewer and Sonar issues.

vincenzo-racca-pa commented on src/main/java/it/pagopa/pn/service/desk/middleware/queue/consumer/handler/ExternalChannelEventHandler.java:
```diff
@@ -1,38 +1,32 @@
-            }
-        };
+    @SqsListener(value = "${pn.service-desk.topics.externalchannel-queue}", acknowledgementMode = SqsListenerAcknowledgementMode.ALWAYS)
+    public void pnExtChannelEventInboundConsumer(Message<SingleStatusUpdateDto> message) {
```
Il default è ON_SUCCESS, quindi o eliminiamo il valore esplicito oppure lo settiamo in ON_SUCESS e non ALWAYS

vincenzo-racca-pa commented on src/main/java/it/pagopa/pn/service/desk/middleware/queue/consumer/handler/PaperChannelEventHandler.java:
```diff
@@ -1,36 +1,38 @@
-            } catch (Exception ex) {
-                log.error("Error in pnPaperChannelInboundConsumer {}", ex.getMessage());
-                throw ex;
+    @SqsListener(value = "${pn.service-desk.topics.paperchannel-queue}", acknowledgementMode = SqsListenerAcknowledgementMode.ALWAYS)
```
uguale

vincenzo-racca-pa commented on src/main/java/it/pagopa/pn/service/desk/middleware/queue/consumer/handler/SafeStorageEventHandler.java:
```diff
@@ -1,42 +1,40 @@
-            }
-        };
+
+    @SqsListener(value = "${pn.service-desk.topics.safe-storage-events}", acknowledgementMode = SqsListenerAcknowledgementMode.ALWAYS)
```
uguale

vincenzo-racca-pa commented on src/main/java/it/pagopa/pn/service/desk/middleware/queue/consumer/handler/ServiceDeskEventHandler.java:
```diff
@@ -1,51 +1,55 @@
-                throw ex;
-            }
-        };
+    @SqsListener(value = "${pn.service-desk.topics.internal-queue}", acknowledgementMode = SqsListenerAcknowledgementMode.ALWAYS)
```
uguale

vincenzo-racca-pa commented on src/main/resources/application.properties:
```diff
@@ -14,11 +14,6 @@ pn.service-desk.event.handler.NOTIFY_DELIVERY_PUSH=notifyDeliveryPushInboundCons
 pn.middleware.init.aws=true
 
-# Queue inbound event configuration
-spring.cloud.stream.function.routing.enabled=true
```
possiamo eliminare anche:

```java
pn.service-desk.event.handler.SERVICE_DESK_INTERNAL_EVENTS=pnInternalEventInboundConsumer
pn.service-desk.event.handler.SAFE_STORAGE_EVENTS=pnSafeStorageEventInboundConsumer
pn.service-desk.event.handler.VALIDATION_OPERATIONS_EVENTS=validationOperationsInboundConsumer
pn.service-desk.event.handler.PAPER_CHANNEL_EVENTS=pnPaperChannelInboundConsumer
pn.service-desk.event.handler.EXTERNAL_CHANNEL_EVENTS=pnExtChannelEventInboundConsumer
pn.service-desk.event.handler.NOTIFY_DELIVERY_PUSH=notifyDeliveryPushInboundConsumer
```

in quanto venivano usati per spring cloud stream. E possiamo quindi eliminare anche la classe `EventHandler`

vincenzo-racca-pa commented on src/test/java/it/pagopa/pn/service/desk/config/BaseTest.java:
```diff
@@ -26,13 +22,12 @@ public abstract class BaseTest {
 
     @Slf4j
     @SpringBootTest
-    @EnableAutoConfiguration(exclude= {SqsAutoConfiguration.class, ContextFunctionCatalogAutoConfiguration.class})
```
Questo non va eliminato va ma modificato in:

```java
@EnableAutoConfiguration(exclude= {SqsAutoConfiguration.class })
```

in quanto la classe WithMockServer è fatta di proposito per non far inizializzare gli oggetti AWS

vincenzo-racca-pa commented on src/test/resources/application.properties:
```diff
@@ -22,7 +22,7 @@ pn.service-desk.external-channel-basepath=http://localhost:${mockserver.bean.por
 pn.service-desk.x-pagopa-extch-cx-id=pn-cons-000
 
-
+spring.cloud.aws.sqs.enabled=false
```
da eliminare questa property. Era necessario evidentemente perché si era eliminato `EnableAutoConfiguration(exclude= {SqsAutoConfiguration.class, ContextFunctionCatalogAutoConfiguration.class})`

vincenzo-racca-pa commented on pom.xml:
```diff
@@ -4,29 +4,42 @@
 			<groupId>it.pagopa.pn</groupId>
 			<artifactId>pn-commons</artifactId>
-			<version>2.2.5</version>
+			<version>2.15.0-SPRINGBOOT3</version>
```
eliminare gli exclusions in quanto nel nuovo commons abbiamo eliminato la dipendenza da lollipop

vincenzo-racca-pa commented on pom.xml:
```diff
@@ -4,29 +4,42 @@
 	</properties>
 
 	<dependencies>
+		<dependency>
```
1. Eliminare la dipendenza: spring-cloud-function-context (questo porterà ad eliminare dalle classi di test `@FunctionalSpringBootTest` ed estendere invece `extends BaseTest` che ha già `@Import(LocalStackTestConfig.class)`, da eliminare quindi nella classe figlia)
2. Eliminare il tag version di tutte le dipendenze con groupId software.amazon.awssdk in quanto la versione deve essere presa dal parent pom.
3. Inoltre vedo dichiarate 2 volte le dipendenze di mockserver-junit-jupiter e mockserver-netty (cancellare quelle con versione più vecchia)

RiccardoPompili committed:
PN-17484 - Notes

michelescara committed:
Merge branch 'develop' into feature/PN-17484

# Conflicts:
#	pom.xml

michelescara committed:
PN-17484: workaround for discriminator handling in OpenAPI Generator 7.4.0

Replicated the workaround already present for version 5.4.0 to fix polymorphic deserialization issue in pn-delivery-push clients.

OpenAPI Generator produces @JsonTypeInfo/@JsonSubTypes annotations on classes with discriminator, but subtypes don't extend the base class, causing Jackson deserialization errors. It seems to be a client side only issue.

Changes:
  - Added custom pojo.mustache template for version 7.4.0
  - Added README.md documenting the workaround
  - Updated 5.4.0 README.md with note about incorrect expectation
  - Added categoryType in mock for PnDeliveryPushClientTest

michelescara committed:
PN-17484: add missing README.md.

vincenzo-racca-pa commented on src/main/java/it/pagopa/pn/service/desk/config/msclient/ExternalChannelApiConfigurator.java:
```diff
@@ -9,7 +9,7 @@
 import org.springframework.context.annotation.Configuration;
 
-import javax.annotation.PostConstruct;
+import jakarta.annotation.PostConstruct;
```
@RiccardoPompili questo import non è utilizzato, lo possiamo eliminare (mi sa che ce ne sono altri in questa classe)

vincenzo-racca-pa commented on src/main/java/it/pagopa/pn/service/desk/service/impl/ApiKeysServiceImpl.java:
```diff
@@ -9,11 +9,13 @@
 import it.pagopa.pn.service.desk.service.ApiKeysService;
 import it.pagopa.pn.service.desk.service.AuditLogService;
 import lombok.CustomLog;
+import org.springframework.http.HttpStatus;
```
@RiccardoPompili anche questo import non utilizzato, lo eliminerei

vincenzo-racca-pa commented on src/main/java/it/pagopa/pn/service/desk/service/impl/InfoPaServiceImpl.java:
```diff
@@ -14,14 +14,19 @@
 import it.pagopa.pn.service.desk.mapper.InfoPaMapper;
+import it.pagopa.pn.service.desk.utility.Utility;
 import lombok.CustomLog;
+import org.springframework.http.HttpStatus;
```
@RiccardoPompili anche qui

vincenzo-racca-pa commented on src/main/java/it/pagopa/pn/service/desk/service/impl/ProfileServiceImpl.java:
```diff
@@ -14,13 +14,15 @@
 import it.pagopa.pn.service.desk.service.ProfileService;
 import lombok.CustomLog;
 import org.jetbrains.annotations.NotNull;
+import org.springframework.http.HttpStatus;
```
@RiccardoPompili idem qui

vincenzo-racca-pa commented on src/test/resources/application.properties:
```diff
@@ -22,8 +22,6 @@ pn.service-desk.external-channel-basepath=http://localhost:${mockserver.bean.por
-
-
 # S3 Bucket
 aws.bucket.name=local-doc-bucket
```
@RiccardoPompili qui manca la modifica che invece è stata fatta correttamente sull'application.properties dentro la cartella config (spring.aws....). Se inevec i test funzionano lo stesso, allora significa che 

aws.profile-name=default
aws.region-code=us-east-1
aws.endpoint-url=http://localhost:4566

dentro il file di test sono superflue, quindi meglio eliminarle.

vincenzo-racca-pa commented on src/main/resources/templates/7.4.0/client/README.md:
```diff
@@ -0,0 +1,37 @@
+# Template personalizzato per OpenAPI Generator 7.4.0
+
+## Contesto
```
@RiccardoPompili non mi torna il workround (vedi pn-delivery-push-workflow che usa come client pn-timeline-service, che ha i discriminator. Le classi client generate su pn-delivery-push-workflow hanno correttamente l'ereditarietà)

RiccardoPompili committed:
fix

michelescara committed:
Merge branch 'develop' into feature/PN-17484

# Conflicts:
#	pom.xml
#	src/test/java/it/pagopa/pn/service/desk/utility/UtilityTest.java

michelescara committed:
PN-17484: Fix critical issues.

vincenzo-racca-pa committed:
Merge branch 'develop' into feature/PN-17484
